sound/soc/sof/intel/hda-loader-skl.c

```diff
@@ -418,3 +445,140 @@ int hda_dsp_cl_boot_firmware_skl(struct snd_sof_dev *sdev)
 	dev_err(sdev->dev, "error: load fw failed err: %d\n", ret);
 	return ret;
 }
+
+int cl_skl_cldma_wait_interruptible(struct snd_sof_dev *sdev)
+{
+	int ret = 0;
+
+	if (!wait_event_timeout(sdev->waitq,
+				sdev->code_loading,
+				msecs_to_jiffies(HDA_SKL_WAIT_TIMEOUT))) {
+		dev_err(sdev->dev, "%s: Wait timeout\n", __func__);
+		ret = -EIO;
+		goto cleanup;
+	}
+
+	dev_dbg(sdev->dev, "%s: Event wake\n", __func__);
+	if (sdev->code_loading != 1) {
+		dev_err(sdev->dev, "%s: DMA Error\n", __func__);
+		ret = -EIO;
+	}
+
+cleanup:
+	sdev->code_loading = 0;
+	return ret;
+}
+
+static void cl_skl_cldma_fill_buffer(struct snd_sof_dev *sdev,
+				     unsigned int bufsize,
+				     unsigned int copysize,
+				     const void *curr_pos,
+				     bool intr_enable, bool trigger)
+{
+	/* 1. copy the image into the buffer with the maximum buffer size. */
+	unsigned int size = (bufsize == copysize) ? bufsize : copysize;
+
+	memcpy(sdev->dmab.area, curr_pos, size);
+
+	/* 2. Setting the wait condition for every load. */
+	sdev->code_loading = 0;
+
+	/* 3. Set the interrupt. */
+	if (intr_enable)
+		cl_skl_cldma_set_intr(sdev, true);
+
+	/* 4. Set the SPB. */
+	cl_skl_cldma_setup_spb(sdev, size, trigger);
+
+	/* 5. Trigger the code loading stream. */
+	if (trigger)
+		cl_skl_cldma_stream_run(sdev, true);
+}
+
+static int
+cl_skl_cldma_copy_to_buf(struct snd_sof_dev *sdev, const void *bin,
+			 u32 total_size, u32 bufsize)
+{
+	int ret = 0;
+	unsigned int bytes_left = total_size;
+	const void *curr_pos = bin;
+
+	if (total_size <= 0)
+		return -EINVAL;
+
+	dev_dbg(sdev->dev, "Total binary size: %u\n", total_size);
+
+	while (bytes_left > 0) {
+		if (bytes_left > bufsize) {
+			cl_skl_cldma_fill_buffer(sdev, bufsize, bufsize,
+						 curr_pos, true, true);
+			ret = cl_skl_cldma_wait_interruptible(sdev);
+			if (ret < 0) {
+				cl_skl_cldma_stream_run(sdev, false);
+				return ret;
+			}
+			bytes_left -= bufsize;
+			curr_pos += bufsize;
+		} else {
+			cl_skl_cldma_set_intr(sdev, false);
+			cl_skl_cldma_fill_buffer(sdev, bufsize, bytes_left,
+						 curr_pos, false, true);
+			return 0;
+		}
+	}
+
+	return bytes_left;
+}
+
+#define STRIP_FIRMWARE		1
```

Comment:
Whats this for, I don't think we need it for SOF ?

Reply:
This is for the offset of the binary, when using cAVS binary for debug, we need set it as true, and SOF don't need this. I will remove it.